File: src/data/courses/dsa/binary-search-trees/bst-core-operations/validate-bst/code/java/ValidateBST.java
public class ValidateBST {
    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int x) { val = x; }
    }
    
    private boolean isValidBST(TreeNode root, long minVal, long maxVal) {
        if (root == null) return true;
        
        if (root.val <= minVal || root.val >= maxVal) return false;
        
        return isValidBST(root.left, minVal, root.val) &&
               isValidBST(root.right, root.val, maxVal);
    }
    
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    
    public static void main(String[] args) {
        TreeNode root = new TreeNode(2);
        root.left = new TreeNode(1);
        root.right = new TreeNode(3);
        
        Example sol = new Example();
        System.out.println("Is Valid BST: " + sol.isValidBST(root));
    }
}

File: src/data/courses/dsa/binary-search-trees/bst-core-operations/validate-bst/code/cpp/validate_bst.cpp
// ============================================================
// Validate BST
// ============================================================
#include <climits>
using namespace std;
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int v):val(v),left(nullptr),right(nullptr){} };

// Solution 1: Range Validation
class Solution1 {
public:
    bool isValidBST(TreeNode* root) {
        return validate(root, LONG_MIN, LONG_MAX);
    }
    bool validate(TreeNode* node, long minVal, long maxVal) {
        if (!node) return true;
        if (node->val <= minVal || node->val >= maxVal) return false;
        return validate(node->left, minVal, node->val) &&
               validate(node->right, node->val, maxVal);
    }
};

// Solution 2: Inorder Check
class Solution2 {
    TreeNode* prev = nullptr;
public:
    bool isValidBST(TreeNode* root) {
        prev = nullptr;
        return inorder(root);
    }
    bool inorder(TreeNode* node) {
        if (!node) return true;
        if (!inorder(node->left)) return false;
        if (prev && node->val <= prev->val) return false;
        prev = node;
        return inorder(node->right);
    }
};
